MM-60480: Do not invalidate channel member cache on login

agnivade opened the pull request:
On user login, we were aggressively clearing the channel member
cache when it wasn't necessary. There is no channel membership
info that is changing on a user login.

Additionally, we fix some more issues with pointer passing
to the cache. It's a known problem that our value passing
style isn't consistent when we fetch items from cache.
Sometimes we pass pointer-to-pointer, sometimes it's just a
pointer. This is also the reason for https://github.com/mattermost/mattermost/pull/27830.

We fix 2 such cases where we passed pointer-to-pointer
but didn't handle the special case inside the cache.
This time, we actually fix it in the app layer instead
of the cache layer and also add a test for good measure.

https://mattermost.atlassian.net/browse/MM-60480

```release-note
NONE
```

agarciamontoro commented on server/channels/store/localcachelayer/channel_layer.go:
```diff
@@ -291,7 +291,7 @@ func (s LocalCacheChannelStore) GetAllChannelMembersForUser(ctx request.CTX, use
 	}
 	if allowFromCache {
-		ids := make(map[string]string)
+		var ids model.StringMap
```
Why do we need this?

agarciamontoro replied:
Got it!

agarciamontoro reviewed (commented):
The removal of the redundant invalidations look good!

Regarding the other change: I still don't fully understand this pointer/pointer-to-pointer differences we have in the cache, to be honest. Would it be possible to sync one day for me to fully understand this? I believe the main problem here may be that I don't really understand our caches in the first place.

agnivade commented:
> Regarding the other change: I still don't fully understand this pointer/pointer-to-pointer differences we have in the cache, to be honest. Would it be possible to sync one day for me to fully understand this? 

A simple example to explain:

```go
package main

import "fmt"

type S1 struct {
}

func (*S1) Unmarshal() {
	fmt.Println("called!")
}

type Unmarshaler interface {
	Unmarshal()
}

func main() {
	var aa S1
	getFromCache(&aa) // works!

	var bb *S1
	getFromCache(&bb) // does not work
}

func getFromCache(obj any) {
	if val, ok := obj.(Unmarshaler); ok {
		val.Unmarshal()
	}
}
```

So the real problem is in our codebase, we mix the `aa` and `bb` pattern like crazy. It's "some", not a huge work to go through all the cases and change `bb` to `aa`. But then, most importantly there should be unit tests to prevent someone from using `bb` again.

streamer45 reviewed (approved):
Nice!

agarciamontoro reviewed (approved):
Thanks for the explanation!

agnivade commented:
/update-branch

DHaussermann commented:
@agnivade 0/5 maybe this is not related to your PR but I see a strange behavior on the `/cloud` server I have created for this.  

When I try to use the user picker for mentions or with the `/invite` command what seems to happen is it will get me the first 20 or so users but then no more. Keystrokes do not work to fetch me any new results. Using the GUI I can search for my 2 test users no problem but they do not appear for me in the user picker.
 
![image](https://github.com/user-attachments/assets/138307fe-3646-43e9-8e1e-47fd57c57876)
![image](https://github.com/user-attachments/assets/4c600434-59e2-4173-99a6-0c3fd457ada6)

The web requests to fetch the data are all coming back as response `200` the problem must be elswhere
![image](https://github.com/user-attachments/assets/01c5ee2a-007d-4fff-9aa4-a7f24251fbe7)

This did not repo on `master` when I tried but in fairness, I feel like I don't  conclusively know exact repo steps so I'm. not sure if this could be related to your PR.

Steps:
- Create a `/cloud` server (in HA) running commit ` 07c23ee`
- Invite 2 new users to the team (_dkh-member01_ and _dkh-member02_)
- Enable test commands and add 48 more users (Their usernames all start with `a...`)
- Try using the @ mention autocomplete or invite command to invite the `dkh` users to various new channels
**Observed**: User picker autocomplete cannot find anybody who is not in the initial list of 20 even though my `dkh` users are team members.

Please advise if you can reproduce or provide addtional isolation steps for my affected server.

agnivade commented:
Thanks Dylan. Will take a look.

agnivade commented:
Tested this in my local env. Could not find any difference between the /invite autocomplete and add members modal. Although the add members modal had some more users not in team, I saw the same behavior in master as well.

![modal](https://github.com/user-attachments/assets/c05b9c9b-b791-4fab-986c-21e4088174c4)
![invite_slash](https://github.com/user-attachments/assets/e8d2e792-98f2-46c9-acd8-83fc53dce4c3)


What I did was:
- added 3 testusers manually to a new team.
- added 20 users starting with `a` with the /test command.

Pressing a showed all the 20 users, and pressing test showed all 3 testusers. 

The key thing to note here is that - the only behavior change in this PR is during login. If you are observing anything weird, I'd suggest to logout, then login again and try it out. It is very strange that this PR should affect user autocomplete in any ways, because all it does is does not invalidate some caches. But searches are not cached in any way.

Dylan - if you can create another spinwick from the GitHub label, then me and you can debug this together. Let me know. Thanks.

DHaussermann commented:
@agnivade I'm creating a Spinwick now.
Another option would be to DM credentials to the affected server here https://dkh-pr-28143.test.mattermost.cloud/

Logout / login did not change the behavior. If it's then unrelated to your change I would try again to repro on `master`. Curious if there might be a way to isolate better what's happening here.

DHaussermann commented:
@agnivade I'm not sure what I'm doing differently or if maybe there's something obvious I have missed here that explains it but, I see the issue on the Spinwick.

I'm super curious if you go open a session with your browser will you see the same thing? I created a channels named `pickles` and cannot invite the `dkh` users using the `/invite` command.

![image](https://github.com/user-attachments/assets/11f861d1-45d5-4b77-96b6-720f5a0bdaa5)

![image](https://github.com/user-attachments/assets/856cbcb2-34c9-45b6-8f39-822051d790a8)

agnivade commented:
Thanks Dylan. I'll take a look.

agnivade commented:
Hey @DHaussermann - this was a weird one indeed!

I added a new user with name `agniva`, that works. Then I thought that `a` might collide with the other users already added. Then I changed the name to `ibrahim`, then `claudio`. All worked. Then I changed to `dylan`. Suddenly it stops working!

After a lot of trial and error, I think I have figured it out. It's to do with the letter `d` :smile: The thing is that all of our search results are limited to 25. After you have added the new users, when you type `d`, your dkh user doesn't appear. But if you test with any other user starting with `i` or `c`, the problem doesn't appear. That's because the entry contains within those first 25 results.

But more importantly, if you actually search for `dkh` using the autocomplete API, the results appear. So I think this is actually a webapp bug somehow where it is only filtering results within the first 25 entries.

```
curl -s 'https://mattermost-pr-28143.test.mattermost.cloud/api/v4/users/autocomplete?in_team=qkmkkm6ybin5bbfjn7h5uzehtw&in_channel=x1qgf9r4mbg6xgf5ukc47rq37y&name=dkh&limit=25'   -H 'accept: */*'   -H 'accept-language: en'   -H 'cache-control: no-cache'   -H 'cookie: MMAUTHTOKEN=kkdyoabazjg1dgdwsaap4khatw; MMUSERID=bthj5kgqi3f5t8m6yci85bgjoa; MMCSRF=168oiuk6fpfj5bqh698ab5jhna'   -H 'dnt: 1'   -H 'pragma: no-cache'   -H 'priority: u=1, i'   -H 'x-requested-with: XMLHttpRequest' | jq . | grep username
      "username": "dkh-member01",
      "username": "dkh-member02",
```

Could you try with some more users starting with `i`, `c` or some other letters? Also, I don't understand why this does not show in master. This feels like a webapp bug to me. But I'd be really surprised if this PR somehow affected some behavior in webapp.

DHaussermann commented:
Ah thanks for investigating @agnivade 
Worth mentioning that on my other `/cloud` server https://dkh-pr-28143.test.mattermost.cloud the "pre-canned" user named `user` was also affected so I was leaning more towards the problem perhaps being it can't fetch additional users not in the initial list.
Regarding not being able to repro on `master` I can't explain but... I can try and create a more conclusive test. I will try to create a `/cloud` server where I'm sure it reproduces, then roll that code forward to `master`.

DHaussermann commented:
@agnivade just to close the thread above with the user picker... This was more difficult to repro than I though (on this branch or `master`) because most of the time, for some reason, keystrokes **will** refine results to go find the user even if they are not in the initial result set when I hit `@` so, even when the issue is present only a relatively small subset of users seem affected. 
I was eventually able to repro on `master` and opened a ticket here https://mattermost.atlassian.net/browse/MM-60636

agnivade commented:
Very nice work Dylan. This was a very tricky bug.

DHaussermann reviewed (approved):
Tested and passed
- Tested adding and removing users from channels where test user is a member
- Ensured channel memberships remain accurate for the test users sessions
- Tested adding and removing members from channels while the test user was logged out and then logging in again as test user to ensure channel membership is accurate and no issues seeing the current members.
- Tested including test user in new GM channels both while they were logged in and logged out. 
No issues found LGTM!

Reported pre-existing issue with user picker separately.  Thanks for the help with this @agnivade

agnivade commented:
/update-branch
